src/github_repo_request_local.py

```diff
@@ -83,45 +164,154 @@ def list_test_files(directory, excluded_extensions):
+
+# Get the path to save the output file
+output_file_path = repo_root / "data" / "repo_testfiles_output.txt"
+
+# Redirect stdout to a file for outputting repo URLs and test file names
+original_stdout = sys.stdout
+
+with open(output_file_path, "a") as f:
+    sys.stdout = f
+
+    for index, row in df.iterrows():
+        # Check if we need to skip this repository because it's fully processed
+        if row["testfilecountlocal"] != -1 and pd.notna(row["last_commit_hash"]):
+            continue
+
+        original_repo_url = row["repourl"]
+        repo_url = get_base_repo_url(original_repo_url)
+        if not repo_url:
+            logger.info(f"Invalid repository URL: {original_repo_url}")
             continue
+        repo_name = Path(repo_url.split("/")[-1]).stem
+        clone_dir = clone_dir_base / repo_name
+
+        # Clone only if directory doesn't exist
+        if not clone_dir.exists():
+            try:
+                logger.info(f"Trying to clone {repo_url} into {clone_dir}")
+                subprocess.run(
+                    ["git", "clone", repo_url, str(clone_dir)],
+                    check=True,
+                    capture_output=True,
+                )
+                logger.info(f"Successfully cloned the repo: {repo_name}")
+
+                # Fetch the last commit hash and store it in the DataFrame
+                last_commit_hash = get_last_commit_hash(clone_dir)
+                if last_commit_hash is not None:
+                    df.at[index, "last_commit_hash"] = last_commit_hash
+
+                # On successful clone, increment the processed_count
+                processed_count += 1
+
+            except subprocess.CalledProcessError as e:
+                logger.error(
+                    f"Failed to clone the repo: {repo_name}." f" " f"Exception: {e}"
+                )
+                df.at[index, "testfilecountlocal"] = -1
+
+                # Even on failure, consider it processed for this batch
+                processed_count += 1
+
+                continue
 
-    test_file_names = list_test_files(clone_dir, args.exclude)
-    count = len(test_file_names)
-    df.at[index, "testfilecountlocal"] = count
-    formatted_names = "\n".join(test_file_names)
-    logger.info(f"Test file names for {repo_name}:\n{formatted_names}")
+        # Always attempt to fetch the last commit hash if not already fetched
+        if pd.isna(row["last_commit_hash"]):
+            last_commit_hash = get_last_commit_hash(clone_dir)
+            df.at[index, "last_commit_hash"] = last_commit_hash
 
-    # Save after each update
+        # Count test files if not already counted
+        if row["testfilecountlocal"] == -1:
+            test_file_names = list_test_files(clone_dir, args.exclude)
+            count = len(test_file_names)
+            df.at[index, "testfilecountlocal"] = count
+            # Output formatted repo URL and test files
+            print(f"Repository URL: {repo_url}")
```

Comment:
Assuming you modify the code to write this info to a file, these 5 or so lines will need modifying to write to a file.

Reply:
- reverted the changes and removed the redirection of the stdout output
 - Implementation of file writing within the repository processing loop.
 - Addition of detailed comments and logging for new functionalities.
 - Update to the script's `docstring` to reflect new capabilities and enhancements.